Source/Core/Runtime/SceneObjects/ProcessSceneObject.cs

```diff
@@ -151,23 +277,92 @@ protected void Init()
+#if UNITY_EDITOR
+        //TODO Possibly move this in to helper class
+        /// <summary>
+        /// Checks if the asset is saved on disk (e.g: a prefab in edit mode).
+        /// </summary>
+        /// <returns><c>true</c> if the asset is saved on disk; otherwise, <c>false</c>.</returns>
+        public bool IsAssetOnDisk()
         {
-            if (RuntimeConfigurator.Exists)
+            // Happens when in prefab mode and adding or removing components
+            if (this == null)
             {
-                RuntimeConfigurator.Configuration.SceneObjectRegistry.Unregister(this);
+                return false;
+            }
+
+            return PrefabUtility.IsPartOfPrefabAsset(this) || IsEditingInPrefabMode();
+        }
+
+        private bool wasInPrefabMode = false;
+
+        /// <summary>
+        /// Determines whether the current object is being edited in prefab mode.
+        /// </summary>
+        /// <returns><c>true</c> if the object is being edited in prefab mode; otherwise, <c>false</c>.</returns>
+        private bool IsEditingInPrefabMode()
+        {
+            if (EditorUtility.IsPersistent(this))
+            {
+                // if the game object is stored on disk, it is a prefab of some kind, despite not returning true for IsPartOfPrefabAsset =/
+                wasInPrefabMode = true;
+                return true;
+            }
+            else
+            {
+                // If the GameObject is not persistent let's determine which stage we are in first because getting Prefab info depends on it
+                var mainStage = StageUtility.GetMainStageHandle();
+                var currentStage = StageUtility.GetStageHandle(gameObject);
+                if (currentStage != mainStage)
+                {
+                    var prefabStage = PrefabStageUtility.GetPrefabStage(gameObject);
+                    if (prefabStage != null)
+                    {
+                        wasInPrefabMode = true;
+                        return true;
+                    }
+                }
+            }
+            if (wasInPrefabMode)
+            {
+                CheckMainStageTransition();
+            }
+            wasInPrefabMode = false;
+            return false;
+        }
+
+        private void CheckMainStageTransition()
+        {
+            //Make sure we are in back in the main scene and not left a nester prefab stage
+            if (StageUtility.GetCurrentStageHandle() == StageUtility.GetMainStageHandle())
+            {
+                //FIXME We need to update the SceneObjectRegistry because we do not do it in prefab edit mode
+                // it is possible that PSO as well as Tags have been added or removed
+                SceneObjectRegistryV2 reg = RuntimeConfigurator.Configuration.SceneObjectRegistry as SceneObjectRegistryV2;
+                reg.DebugRebuild();
+                //TODO wee need to set the current selected competent dirty if its a PSO
```

Comment:
@VaLiuM09 I would like to have you input here on how to solve this as it is connected to the SceneObjectRegistry.

Reply:
I would not call DebugRebuild which is a rough function that clears the registry and registers everything from scratch. If we need to, we could add a function to the registry interface that does a sanity check of the registry by removing all invalid object references, then call RegisterAll as we know nothing happens if the object is already registered. Does it make sense?